src/rust/engine/rule_graph/src/builder.rs

```diff
@@ -383,6 +383,37 @@ impl<'t, R: Rule> Builder<'t, R> {
       to_visit.into_iter().rev().collect::<IndexSet<_>>()
     };
 
+    // Both the in_set and out_set shrink as this phase continues and nodes are monomorphized.
+    // Because the sets only shrink, we are able to eliminate direct dependencies based on their
+    // absence from the sets.
+    //
+    // If a node has been monomorphized and all of its dependencies have minimal in_sets, then we
+    // can assume that its in_set is minimal too: it has _stopped_ shrinking. That means that we can
+    // additionally prune dependencies transitively in cases where in_sets contain things that are
+    // not in a node's out_set (since the out_set will not grow, and the minimal in_set represents
+    // the node's true requirements).
+    let mut minimal_in_set = HashSet::new();
+
+    // Should be called after a Node has been successfully reduced (regardless of whether it became
+    // monomorphic) to maybe mark it minimal.
+    let maybe_mark_minimal_in_set = |minimal_in_set: &mut HashSet<NodeIndex<u32>>,
```

Comment:
I don't think this closure is actually closing over any variables here, so why not define it as a function scoped within the outer `monomorphize` function?